Source/santad/EventProviders/EndpointSecurity/Message.mm

```diff
@@ -38,13 +41,36 @@
   other.es_msg_ = nullptr;
   process_token_ = std::move(other.process_token_);
   other.process_token_ = std::nullopt;
+  stat_change_step_ = other.stat_change_step_;
+  stat_error_ = other.stat_error_;
 }
 
 Message::Message(const Message &other) {
   esapi_ = other.esapi_;
   es_msg_ = other.es_msg_;
   esapi_->RetainMessage(es_msg_);
   process_token_ = other.process_token_;
+  stat_change_step_ = other.stat_change_step_;
+  stat_error_ = other.stat_error_;
+}
+
+void Message::UpdateStatState(santa::santad::StatChangeStep step) const {
+  // Only update state for AUTH EXEC events and if no previous change was detected
+  if (es_msg_->event_type == ES_EVENT_TYPE_AUTH_EXEC &&
+      stat_change_step_ == santa::santad::StatChangeStep::kNoChange &&
+      // Note: The following checks are required due to tests that only
+      // partially construct an es_message_t.
+      es_msg_->event.exec.target && es_msg_->event.exec.target->executable) {
+    struct stat &es_sb = es_msg_->event.exec.target->executable->stat;
+    struct stat sb;
+    errno = 0;
+    int ret = stat(es_msg_->event.exec.target->executable->path.data, &sb);
+    // If stat failed, or if devno/inode changed, update state.
+    if (ret != 0 || es_sb.st_ino != sb.st_ino || es_sb.st_dev != sb.st_dev) {
+      stat_change_step_ = step;
+      stat_error_ = errno;
```

Comment:
It would be helpful to tell a part a stat error, inode mismatch and dev mismatch. Currently error is 0 for both inode and dev mismatches. We don't want to overload errno with additional meaning. I can't think of a good approach besides splitting up these kinds of failures or adding a new dimension to the metric. Both are not great.

How concerned are we about stat failing? Maybe just don't log a metric if stat fails and only log mismatches, which is what we really care about?

Reply:
My hunch would be that the biggest reason stat fails would be due to the file being deleted. This feels like that would be interesting information to capture..

But I could see grouping all stat errors together. Then we could change `stat_error_` to an enum type. Something like:

```
enum class StatResult {
  kOK,
  kStatError,
  kInodeDevnoMismatch,
}
```

So then it would just boil down to how much would we care to know the underlying error if we see a lot of them...

Reply:
StatResult sounds good to me